### PR TITLE
Another pacing algorithm

### LINKED ISSUE #3122: Recommend algorithm for packet scheduler, rather than implementation
The current text in recovery says:

> As an example of a well-known and publicly available implementation of a flow pacer, implementers are referred to the Fair Queue packet scheduler (fq qdisc) in Linux (3.11 onwards).

Could we change this recommendation to something like "a Fair Queue packet scheduler using an algorithm such as "Weighted Round Robin" or "Deficit Round Robin"? The specific Linux reference is not very helpful in online searches.

### PR DESCRIPTION
**Pull request by ianswett**
This is based on @martinthomson #3630, but my suggestions were so large I created a new PR.

Fixes #3122

### CODE REVIEW

File: draft-ietf-quic-recovery.md [hunk 1]
```diff
@@ -906,9 +906,26 @@ congestion window, which is recommended to be the minimum of
 max_datagram_size is the current maximum size of a datagram for the connection,
 not including UDP or IP overhead.
 
-As an example of a well-known and publicly available implementation of a flow
-pacer, implementers are referred to the Fair Queue packet scheduler (fq qdisc)
-in Linux (3.11 onwards).
+Endpoints can implement pacing as they choose, as long as bursts are not larger
+than the initial window and average sending rate over a round trip does not
+exceed a small multiple, `pacing_gain`, of the congestion window. Larger
+bursts can reduce CPU utilization at the risk of increased retransmits rates.
```
Comment (martinthomson) [inline draft-ietf-quic-recovery.md:912]: This last sentence seems a little disconnected from the rest.  A small tweak might might help:

```suggestion
exceed a small multiple, `pacing_gain`, of the congestion window. Endpoints can
trade off sending packets in bursts, which might be done to reduce CPU
utilization, against even pacing, which has a lower risk of inducing packet
loss.
```

File: draft-ietf-quic-recovery.md [hunk 2]
```diff
@@ -906,9 +906,26 @@ congestion window, which is recommended to be the minimum of
...
+
+A `pacing_gain` multiplier greater than 1 allows senders to increase the rate
+of sending. The value of `pacing_gain` might be larger in slow start than
+congestion avoidance to ensure the sender is able to double its congestion
+window every round trip.
```
Comment (martinthomson) [inline draft-ietf-quic-recovery.md:917]: There are two reasons.  As the window increases, the rate increases, so you only really need this gain to a) anticipate future increases and b) avoid having scheduling delays eat into your send rate.

The future increases thing is worth noting.  Pacing causes packets to be spaced out over time, and you won't be able to realize an increase in congestion window until an entire round trip has passed if pacing_gain is 1.  Whereas if pacing_gain is matched to the rate of increase in the window, you are already sending at the increased rate, so you never have to worry about the pacer being the cause of you not sending enough.

@janaiyengar argued that you don't need to worry about this effect.  I think that it's worth compensating for, at least partially.  With pacing, increases to the window will be continuous, or at least multiple times per round trip time, so the degree to which your pacer drags your send rate down is based on the rate of updates (i.e., ACKs) you get.

This is a stair-step function bounded by a straight line (well, a logarithm, but just change conceptually to a log vertical scale).  The steps happen as you get ACKs (or timer pops), so the amount you shade is bounded by how often the updates are.  You get closer to shading the area under the line if you update more often.  The realization here is that you can shade a similar area if you are willing to shade above the line :)  If you set pacing_gain == multiplicative increase, you are not keeping the steps below the diagonal, you are putting them on top by assuming that you will get an increase.  Choosing a value between 1 and m hedges a little.

Comment (janaiyengar) [reply draft-ietf-quic-recovery.md:917]: To clarify what I said: pacing rate is cwnd/rtt (approximately, for window-based senders). In this world, on each ack, the cwnd increases, and your pacing rate therefore increases. At the end of the first RTT in slow start, your window is 2x the IW, and your pacing rate is 2 x IW / rtt. That is as it should be. Now, to compensate for border conditions, it's a good idea to use a multiplier, and that does not need to be any more than a small delta (say 1.25), and it doesn't need to be higher for slow start.

At any rate, this is way more than you need to be saying in this definition.

File: draft-ietf-quic-recovery.md [hunk 3]
```diff
@@ -906,9 +906,26 @@ congestion window, which is recommended to be the minimum of
...
+
+The ideal interval between the previously send packet burst and the next is:
+
+~~~
+interval = smoothed_rtt * bytes_last_sent / congestion_window / pacing_gain
+~~~
+
+Timers are not ideal, so intervals will need to be adjusted to ensure the
+average pacing rate is maintained.  Not doing so will result in a decrease
+in average sending rate when timers consistently fire late.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md:927]: I think what you want to say is that timers usually run with some granularity. I _think_ what you're trying to say is that timers might fire late and that future computations will need to account for that.

**Pull request closed without merging**